### PR TITLE
Add PyTest for package #2

### PR DESCRIPTION
**Pull request by Pytlicek**
- add pytest tests
- refactor the main function where tests were running while loading the package

All tests are passing on my side. Test/Run with:  `python3 -m pytest -v tests/`

Screenshot 2022-10-03 at 22 10 42

### CODE REVIEW

File: src/qth_locator.py [hunk 1]
```diff
@@ -64,24 +54,28 @@ def square_to_location(qth_locator):
...
-    assert -90.0 <= lat <= 90.0
-    assert -180.0 <= lon <= 180.0
+    try:
+        assert isinstance(lat, float)
+        assert isinstance(lon, float)
+        assert -90.0 <= lat <= 90.0
+        assert -180.0 <= lon <= 180.0
+    except AssertionError as e:
+        # print(f"Latitude or Longitude is wrong\n{e}")
+        return None
```
Comment (tomtomecek) [inline src/qth_locator.py]: I would let the code explode on validation error instead. It makes sense to alert on input errors rather than return None in my opinion. There is also lesser to no point in throwing an exception and catching it one line below, because then you can rather use `if/else` and some guard statement.

The exception can help you to build API error handler too. So if one function bombs and the API can propagate that error towards response to the user as BadRequest.

File: tests/test_1_location_to_square.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        (None, (0, 0)),
+        (None, (32, 26)),
+        ("KM32jn", (32.541667, 26.75)),
+        ("KM32jn", (32.570831, 26.750003)),
+        ("KM72kk", (32.437487, 34.875015)),
+        ("JN45fo", (45.603333, 8.456667)),
+        ("KO92so", (52.601484, 39.565160)),
+        ("KM72jb", (32.071209, 34.780089)),
+        ("KM72jb", (32.07, 34.78)),
+    ]
```
Comment (tomtomecek) [inline tests/test_1_location_to_square.py]: Why there is a need for 9 square cases?

I would recommend to build 1 important happy case, or several important ones when conditions are changing, and then around edge cases.

File: tests/test_1_location_to_square.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    for sq, loc in squares:
+        qth = location_to_square(loc[0], loc[1])
+        # print(f'{loc}: {sq}, calculated: {qth}, match:{sq==qth}')
```
Comment (tomtomecek) [inline tests/test_1_location_to_square.py]: You should avoid debugging comments in final code, event test code. For feedback, you have test suite and its errors.

File: src/qth_locator.py [hunk 4]
```diff
@@ -64,24 +54,28 @@ def square_to_location(qth_locator):
     lon += 5.0 / 60 * lon_sub_sq
     lat += 2.5 / 60 * lat_sub_sq
 
-    lon += 0.5 / 60 * lon_ext_sq
-    lat += 0.25 / 60 * lat_ext_sq
+    lat = round(lat, 6)
+    lon = round(lon, 6)
```
Comment (tomtomecek) [inline src/qth_locator.py:52]: 👍

**Review by tomtomecek — CHANGES_REQUESTED** [review]
I see this PR as a great improvement, because of added unit tests!

File: src/qth_locator.py (hunk 1 above)
Comment (Pytlicek) [reply src/qth_locator.py]: Good catch 👍 
Fixed in: https://github.com/Pytlicek/hamradio_functions/pull/8/commits/c50dc3120a26683bd073ad94f904699bee7e5120

File: tests/test_1_location_to_square.py (hunk 2 above)
Comment (Pytlicek) [reply tests/test_1_location_to_square.py]: Yes, we want more locations to test, because of the nature of how they are counted.
1 Square can hold many locations (lat/long combinations) basically square is not an exact location, It is a map square with specific width and height. In another hand, one location should be in one specific square.

File: tests/test_1_location_to_square.py (hunk 3 above)
Comment (Pytlicek) [reply tests/test_1_location_to_square.py]: Good catch 👍 Thanks 🙏 
Fixed in: https://github.com/Pytlicek/hamradio_functions/pull/8/commits/c50dc3120a26683bd073ad94f904699bee7e5120

File: src/qth_locator.py (hunk 1 above)
Comment (Pytlicek) [reply src/qth_locator.py]: New fix here: https://github.com/Pytlicek/hamradio_functions/commit/ee6ee079dcf419d2210c6d534b41df59fb5419be

File: tests/test_1_location_to_square.py (hunk 3 above)
Comment (Pytlicek) [reply tests/test_1_location_to_square.py]: New fix here: https://github.com/Pytlicek/hamradio_functions/commit/ee6ee079dcf419d2210c6d534b41df59fb5419be

File: src/qth_locator.py [hunk 5]
```diff
@@ -64,22 +48,22 @@ def square_to_location(qth_locator):
...
-    lon += 0.5 / 60 * lon_ext_sq
-    lat += 0.25 / 60 * lat_ext_sq
+    lat = round(lat, 6)
+    lon = round(lon, 6)
 
     return lat, lon
 
 
-def location_to_square(lat, lon):
+def location_to_square(lat: float, lon: float) -> str:
```
Comment (tomtomecek) [inline src/qth_locator.py:57]: I like these annotations. 👍

**Review by tomtomecek — APPROVED** [review]
